### PR TITLE
Ignore duplicate claims

### PR DESCRIPTION
**Pull request by g4s8**
#1489 - do not add duplicate claims to `claims.xml`

### CODE REVIEW

**Comment by 0crat** [comment]
Job [#1490](https://github.com/zerocracy/farm/issues/1490) is now in scope, role is `REV`

**Comment by 0crat** [comment]
This pull request [#1490](https://github.com/zerocracy/farm/issues/1490) is assigned to @krzyk[/z](https://www.0crat.com/u/krzyk), here is [why](https://www.0crat.com/footprint/C3NDPUA8L/2080837339); the budget is 15 minutes, see [§4](http://www.zerocracy.com/policy.html#4); please, read [§27](http://www.zerocracy.com/policy.html#27) and when you decide to accept the changes, inform @g4s8[/z](https://www.0crat.com/u/g4s8) (the architect) right in this ticket; if you decide that this PR should not be accepted ever, also inform the architect; this [blog post](http://www.yegor256.com/2015/02/09/serious-code-reviewer.html) will help you understand what is expected from a code reviewer; there will be a monetary reward for this job

File: src/main/java/com/zerocracy/pm/ClaimsItem.java [hunk 1]
```diff
@@ -87,32 +90,45 @@ public void add(final XML claim) throws IOException {
...
-            );
-            final Collection<String> signatures = new Sorted<>(
-                new Mapped<>(
-                    input -> ClaimsItem.signature(new ClaimIn(input)),
-                    new Xocument(item).nodes("/claims/claim")
+            final String signature;
+            try {
+                signature = ClaimsItem.signature(
+                    new ClaimIn(
+                        new XMLDocument(new Xembler(claim).dom())
```
Comment (krzyk) [inline src/main/java/com/zerocracy/pm/ClaimsItem.java]: @g4s8 chaining test fails here (`at com.zerocracy.pm.ClaimOutTest.chainsThem(ClaimOutTest.java:41)`)

**Comment by 0crat** [comment]
Job `gh:zerocracy/farm#1490` already assigned to @krzyk, can't assign to @krzyk

**Comment by 0crat** [comment]
Job `gh:zerocracy/farm#1490` already assigned to @krzyk, can't assign to @krzyk

File: src/main/java/com/zerocracy/pm/ClaimsItem.java (hunk 1 above)
Comment (krzyk) [reply src/main/java/com/zerocracy/pm/ClaimsItem.java]: @g4s8 I think you need to append it into `/claims`, right now it tries to create two `/claim` entries -> xml with two parent nodes

File: src/main/java/com/zerocracy/pm/ClaimsItem.java [hunk 2]
```diff
@@ -87,33 +91,48 @@ public void add(final XML claim) throws IOException {
...
-                new Directives().xpath("/claims").append(claim)
-            );
-            final Collection<String> signatures = new Sorted<>(
+            final List<XML> claims = new XMLDocument(
+                new Xembler(
+                    new Directives()
+                        .add("claims")
+                        .append(claim)
+                ).dom()
+            ).nodes("/claims/claim");
```
Comment (krzyk) [inline src/main/java/com/zerocracy/pm/ClaimsItem.java]: @g4s8 you'll need to add PMD exclusion for DuplicateLiterals

**Comment by g4s8** [comment]
@krzyk thanks, should be fixed in https://github.com/zerocracy/farm/pull/1490/commits/ccd3ec6a8248be7323975022af48206f0f4ab7d6

File: src/main/java/com/zerocracy/pm/ClaimsItem.java [hunk 3]
```diff
@@ -87,33 +91,48 @@ public void add(final XML claim) throws IOException {
...
+            final boolean duplicates = new IoCheckedScalar<>(
+                new Or(
+                    signatures::contains,
+                    new Mapped<>(
+                        xml -> ClaimsItem.signature(new ClaimIn(xml)),
+                        claims
+                    )
+                )
+            ).value();
+            if (duplicates) {
```
[the remark this answers is not in the document]
Comment (g4s8) [reply src/main/java/com/zerocracy/pm/ClaimsItem.java]: @krzyk right, changed the logic to ignore only duplicates

**Comment by 0crat** [comment]
Job was finished in 2 hours, bonus for fast delivery is possible (see [§36](http://www.zerocracy.com/policy.html#36))

**Comment by 0crat** [comment]
@ypshenychka[/z](https://www.0crat.com/u/ypshenychka) please review this job completed by @krzyk[/z](https://www.0crat.com/u/krzyk), as in [§30](http://www.zerocracy.com/policy.html#30); the job will be fully closed and all payments will be made when the quality review is completed

**Comment by 0crat** [comment]
The job [#1490](https://github.com/zerocracy/farm/issues/1490) is now out of scope

**Comment by 0crat** [comment]
Job [#1490](https://github.com/zerocracy/farm/issues/1490) is pending quality review, can't start a new order

**Comment by 0crat** [comment]
Payment to `ARC` for a closed pull request, as in [§28](http://www.zerocracy.com/policy.html#28): +10 point(s) just awarded to @g4s8[/z](https://www.0crat.com/u/g4s8)

**Comment by ypshenychka** [comment]
@0crat quality good

**Comment by 0crat** [comment]
Order was finished, quality is "good": +25 point(s) just awarded to @krzyk[/z](https://www.0crat.com/u/krzyk)

**Comment by 0crat** [comment]
Quality review completed: +8 point(s) just awarded to @ypshenychka[/z](https://www.0crat.com/u/ypshenychka)

**Pull request merged**